webstruct/webannotator.py

```diff
@@ -280,11 +388,47 @@ def to_webannotator(tree, entity_colors=None, url=None):
     >>> wa_trees = [to_webannotator(tree, entity_colors) for tree in trees]  # doctest: +SKIP
 
     """
-    handler = _WaContentHandler(entity_colors)
-    lxml.sax.saxify(tree, handler)
-    tree = handler.out.etree
-    _copy_title(tree)
-    _add_wacolor_elements(tree, handler.entity_colors)
+    if not entity_colors:
+        entity_colors = EntityColors()
+
+    root = deepcopy(tree)
+
+    # We walk the DOM tree in depth first order and number all nodes.
+    # Also we number each text node as first child and tail node as last child.
+    # So when we have start tag in node with number n
+    # and stop tag in node with number n+m,
+    # we should annotate all nodes with numbers n+i, where i in range [1,m).
+    # All these nodes are located on the rigth and below the start
+    # and on the left and below the end.
+    starts, ends = _find_tag_limits(root)
+
+    if len(ends) != len(starts):
+        raise ValueError('len(ends) != len(starts)')
+
+    ordered = _enumerate_nodes_in_dfs_order(root)
+    starts = [s for s in _translate_to_dfs(starts, ordered)]
+    ends = [e for e in _translate_to_dfs(ends, ordered)]
+
+    starts.sort(key=lambda t: (t.dfs_number, t.position))
+    ends.sort(key=lambda t: (t.dfs_number, t.position))
+
+    dfs_order = (max(n for n in ordered.values()) + 1) * [None]
```

Comment:
why not just `max(ordered.values())`?

Reply:
It is not first variant. Such form is a relic of previous versions.